### PR TITLE
Adding better support for hashmaps

### PR DESCRIPTION
**Pull request by limemloh**
## Purpose

Support using `Address` in a `BTreeMap` and contract schemas.
Types implement Hash and add support for hashmaps in `no-std`.

## Changes

- Derive `PartialOrd` and `Ord` for `ContractAddress` and `Address`.
- Implement `SchemaType` for `Address`.
- Add `HashMap` and `HashSet` from the `hashbrown` crate to support `no-std`.
- Make `HashMap` and `HashSet` default to the `fnv` hasher.
- Derive `Hash` for `Amount`, `Timestamp`, `Duration`, `AccountAddress`, `ContractAddress`, `Address`, `ContractName`, `OwnedContractName`, `ReceiveName` and `OwnedReceiveName`.

## Checklist

- [x] My code follows the style of this project.
- [x] The code compiles without warnings.
- [x] I have performed a self-review of the changes.
- [x] I have documented my code, in particular the intent of the
      hard-to-understand areas.
- [x] (If necessary) I have updated the CHANGELOG.

### CODE REVIEW

File: src/types.rs [hunk 1]
```diff
@@ -2,9 +2,19 @@ use crate::constants;
 #[cfg(not(feature = "std"))]
 use alloc::{string::String, string::ToString, vec::Vec};
 #[cfg(not(feature = "std"))]
-use core::{convert, fmt, iter, ops, str};
+use core::{convert, fmt, hash, iter, ops, str};
 #[cfg(feature = "std")]
-use std::{convert, fmt, iter, ops, str};
+use std::{convert, fmt, hash, iter, ops, str};
+
```
Comment (abizjak) [inline src/types.rs:8]: Empty line.

File: src/types.rs [hunk 2]
```diff
@@ -600,14 +610,14 @@ pub struct ContractAddress {
     serde(tag = "type", content = "address", rename_all = "lowercase")
 )]
 #[cfg_attr(feature = "fuzz", derive(Arbitrary))]
-#[derive(PartialEq, Eq, Copy, Clone, Debug)]
+#[derive(PartialEq, Eq, Copy, Clone, PartialOrd, Ord, Debug, Hash)]
```
Comment (abizjak) [inline src/types.rs:613]: Did you check what the generated comparison function is and whether we can make one that generates smaller code?

File: src/schema.rs [hunk 3]
```diff
@@ -176,6 +178,14 @@ impl SchemaType for AccountAddress {
 impl SchemaType for ContractAddress {
     fn get_type() -> Type { Type::ContractAddress }
 }
+impl SchemaType for Address {
+    fn get_type() -> Type {
+        Type::Enum(Vec::from([
+            (String::from("account"), Fields::Unnamed(Vec::from([Type::AccountAddress]))),
+            (String::from("contract"), Fields::Unnamed(Vec::from([Type::ContractAddress]))),
```
[the remark this answers is not in the document]
Comment (limemloh) [reply src/schema.rs:185]: You mean the one where we have `{"type": "contract", "address": {...}}` for contracts and  `{"type": "account", "address": "...." }` for accounts? 
If so, then we would have to extend the contract schema to support this.

[the remark this answers is not in the document]
Comment (abizjak) [reply src/schema.rs]: Yes. It would look out of place as well, since we use a different method for encoding enums in JSON.

File: src/impls.rs [hunk 4]
```diff
@@ -1,11 +1,14 @@
 use crate::{constants::*, traits::*, types::*};
 
 #[cfg(not(feature = "std"))]
-use alloc::{boxed::Box, collections::*, string::String, vec::Vec};
+use alloc::{boxed::Box, collections, string::String, vec::Vec};
```
[the remark this answers is not in the document]
Comment (limemloh) [reply src/impls.rs:4]: No, I don't think there are any changes to whatever is public from this.

File: Cargo.toml [hunk 5]
```diff
@@ -14,6 +14,14 @@ readme = "README.md"
 [dependencies]
 arbitrary = { version = "0.4", features = ["derive"], optional = true }
 
+[dependencies.hashbrown]
+version = "0.11.2"
+default-features = false
+
+[dependencies.fnv]
+version = "1.0.7"
```
[the remark this answers is not in the document]
Comment (limemloh) [reply Cargo.toml]: I have removed the patch version

File: src/impls.rs [hunk 6]
```diff
@@ -467,6 +469,35 @@ pub fn deserial_map_no_length_no_order_check<R: Read, K: Deserial + Ord, V: Dese
     Ok(out)
 }
 
+/// Write a HashMap as a list of key-value pairs in to particular order, without
+/// the length information.
+pub fn serial_hashmap_no_length<'a, W: Write, K: Serial + 'a, V: Serial + 'a>(
```
Comment (abizjak) [inline src/impls.rs:474]: What is this lifetime here for? I can't see it used for any of the arguments.

File: src/impls.rs [hunk 7]
```diff
@@ -499,6 +530,33 @@ pub fn deserial_set_no_length<R: Read, K: Deserial + Ord + Copy>(
     Ok(out)
 }
 
+/// Write a [HashSet](https://doc.rust-lang.org/std/collections/struct.HashSet.html) as a list of keys in no particular order, without the length information.
+pub fn serial_hashset_no_length<'a, W: Write, K: Serial + 'a>(
```
Comment (abizjak) [inline src/impls.rs:534]: Same question about the lifetime.

File: src/impls.rs [hunk 8]
```diff
@@ -565,6 +623,32 @@ impl<K: Deserial + Ord, V: Deserial> Deserial for BTreeMap<K, V> {
...
+/// They are serialized in no particular order.
+impl<K: Serial, V: Serial> Serial for HashMap<K, V> {
+    fn serial<W: Write>(&self, out: &mut W) -> Result<(), W::Err> {
+        let len = self.len() as u32;
+        len.serial(out)?;
+        serial_hashmap_no_length(self, out)
+    }
+}
+
+/// The deserialization of maps assumes their size as a u32.
```
Comment (abizjak) [inline src/impls.rs:637]:
```suggestion
/// The deserialization of maps assumes their size is a u32.
```

File: src/types.rs [hunk 9]
```diff
@@ -600,14 +609,58 @@ pub struct ContractAddress {
...
+#[derive(Eq, Copy, Clone, Debug)]
 pub enum Address {
     Account(AccountAddress),
     Contract(ContractAddress),
 }
 
+// This trait is implemented manually to save bytes
+impl PartialEq for Address {
+    fn eq(&self, other: &Address) -> bool {
+        match (self, other) {
```
Comment (abizjak) [inline src/types.rs:621]: Is there a reason for not pattern matching on `(&self, other)` here?

Is this better or does it not make a difference.

File: src/types.rs [hunk 10]
```diff
@@ -600,14 +609,58 @@ pub struct ContractAddress {
...
 )]
 #[cfg_attr(feature = "fuzz", derive(Arbitrary))]
-#[derive(PartialEq, Eq, Copy, Clone, Debug)]
+#[derive(Eq, Copy, Clone, Debug)]
 pub enum Address {
     Account(AccountAddress),
     Contract(ContractAddress),
 }
 
+// This trait is implemented manually to save bytes
```
Comment (abizjak) [inline src/types.rs:618]: Maybe rephrase this to something like 
"The trait is implemented manually so that the generated Wasm code is smaller."

File: src/types.rs [hunk 11]
```diff
@@ -600,14 +609,58 @@ pub struct ContractAddress {
...
+    fn eq(&self, other: &Address) -> bool {
+        match (self, other) {
+            (Address::Account(s), Address::Account(o)) => s == o,
+            (Address::Contract(s), Address::Contract(o)) => s == o,
+            _ => false,
+        }
+    }
+}
+
+// This trait is implemented manually to save bytes
```
Comment (abizjak) [inline src/types.rs:629]: And here.

File: src/types.rs [hunk 12]
```diff
@@ -600,14 +609,58 @@ pub struct ContractAddress {
...
+            }
+            Address::Contract(address) => {
+                1u8.hash(state);
+                address.hash(state);
+            }
+        }
+    }
+}
+
+// This trait is implemented manually to save bytes
```
Comment (abizjak) [inline src/types.rs:645]: And here.

File: src/types.rs [hunk 13]
```diff
@@ -600,14 +609,58 @@ pub struct ContractAddress {
...
+impl PartialOrd for Address {
+    fn partial_cmp(&self, other: &Address) -> Option<Ordering> { Some(self.cmp(other)) }
+}
+
+// This trait is implemented manually to save bytes
+impl Ord for Address {
+    fn cmp(&self, other: &Address) -> Ordering {
+        match (self, other) {
```
Comment (abizjak) [inline src/types.rs:653]: Same question about `(&self, other)`.

File: src/impls.rs (hunk 6 above)
Comment (limemloh) [reply src/impls.rs:474]: I copied the signature from serial_map_no_length which also have this, but I'll make sure to remove these from both

File: src/types.rs (hunk 9 above)
Comment (limemloh) [reply src/types.rs:621]: I am not sure what you mean by this.
Would you mind elaborating?

Comment (abizjak) [reply src/types.rs:621]: Never mind this, I read it wrong.

File: src/types.rs (hunk 13 above)
Comment (abizjak) [reply src/types.rs:653]: And the same here.

File: src/impls.rs (hunk 6 above)
Comment (abizjak) [reply src/impls.rs:474]: It might have made sense at some point.

**Review by abizjak — DISMISSED** [review]
If you would please fix the formatting, this is good to go.

**Review by abizjak — APPROVED** [review]
